integration/duckdb/src/lance/duckdb/vector_functions.cc

```diff
@@ -0,0 +1,104 @@
+    assert(first.size() == second.size());
+    // TODO: vectorize this computation later
+    // To vectorize the compute, we need either have arrow Functions `Pow(list[float],
+    // list[float])`, `Sub(list[float], list[float])`, or the equivalent in DuckDB.
+    float sum = 0;
+    for (size_t j = 0; j < first.size(); j++) {
+      sum += pow((first[j].GetValue<T>() - second[j].GetValue<T>()), 2);
+    }
+    result.SetValue(i, sum);
+  }
+}
+
+void IsInRectangle(::duckdb::DataChunk &args,
+                   ::duckdb::ExpressionState &state,
+                   ::duckdb::Vector &result) {
+  result.SetVectorType(::duckdb::VectorType::FLAT_VECTOR);
+  for (::duckdb::idx_t i = 0; i < args.size(); i++) {
+    auto point = ::duckdb::ListValue::GetChildren(args.data[0].GetValue(i));
+    auto x = point[0];
+    auto y = point[1];
+
+    auto second = ::duckdb::ListValue::GetChildren(args.data[1].GetValue(i));
+    auto upper_left = ::duckdb::ListValue::GetChildren(second[0]);
+    auto lower_right = ::duckdb::ListValue::GetChildren(second[1]);
+
+    result.SetValue(i,
+                    ((upper_left[0] <= x) && (x <= lower_right[0]) && (upper_left[1] <= y) &&
+                     (y <= lower_right[1])));
+  }
+}
+
+std::vector<std::unique_ptr<::duckdb::CreateFunctionInfo>> GetVectorFunctions() {
+  std::vector<std::unique_ptr<::duckdb::CreateFunctionInfo>> functions;
+
+  ::duckdb::ScalarFunctionSet l2_distance("l2_distance");
+  l2_distance.AddFunction(
+      ::duckdb::ScalarFunction({::duckdb::LogicalType::LIST(::duckdb::LogicalType::INTEGER),
+                                ::duckdb::LogicalType::LIST(::duckdb::LogicalType::INTEGER)},
+                               ::duckdb::LogicalType::INTEGER,
+                               L2Distance<int>));
+  l2_distance.AddFunction(
+      ::duckdb::ScalarFunction({::duckdb::LogicalType::LIST(::duckdb::LogicalType::BIGINT),
+                                ::duckdb::LogicalType::LIST(::duckdb::LogicalType::BIGINT)},
+                               ::duckdb::LogicalType::BIGINT,
+                               L2Distance<int64_t>));
+  l2_distance.AddFunction(
+      ::duckdb::ScalarFunction({::duckdb::LogicalType::LIST(::duckdb::LogicalType::FLOAT),
+                                ::duckdb::LogicalType::LIST(::duckdb::LogicalType::FLOAT)},
+                               ::duckdb::LogicalType::FLOAT,
+                               L2Distance<float>));
+  l2_distance.AddFunction(
+      ::duckdb::ScalarFunction({::duckdb::LogicalType::LIST(::duckdb::LogicalType::DOUBLE),
+                                ::duckdb::LogicalType::LIST(::duckdb::LogicalType::DOUBLE)},
+                               ::duckdb::LogicalType::DOUBLE,
+                               L2Distance<double>));
+  functions.emplace_back(std::make_unique<::duckdb::CreateScalarFunctionInfo>(l2_distance));
+
+  ::duckdb::ScalarFunctionSet rect_contains("rect_contains");
+  rect_contains.AddFunction(::duckdb::ScalarFunction(
```

Comment:
let's also support int/long rectangle regions? i.e., `SELECT rect_contains(tsne, [[0, 0], [300, 300]] FROM dataset`

Reply:
and let's either:

1. change the order of the arguments (usually `rect_contains(a, b)` is interpreted as "the rectangle a contains b", which is the reverse

OR

2. change the name of the function to `in_rectangle`, so it's `a is in rectangle b`

Reply:
ok, changed it to `in_rectangle` for clarity.